frontend/gateway/client/result.go

```diff
@@ -38,6 +42,17 @@ func (r *Result) AddRef(k string, ref Reference) {
 	r.mu.Unlock()
 }
 
+func (r *Result) AddInTotoAttestation(k string, v *attestation.InTotoAttestation, predicateRef Reference) {
+	r.mu.Lock()
+	if r.Attestations == nil {
+		r.Attestations = map[string][]attestation.Attestation{}
+	}
+	v.PredicateRefKey = fmt.Sprintf("attestation:%s", identity.NewID())
+	r.Refs[v.PredicateRefKey] = predicateRef
+	r.Attestations[k] = append(r.Attestations[k], v)
+	r.mu.Unlock()
```

Comment:
I think this function refactor is sub-optimal - it has to be changed from `AddAttestation` to a more specific form, so as to allow passing the `predicateRef` (keeping in mind that other attestation types in the future might require no/multiple refs, so just taking a single one is probably not correct). Not quite sure if I'm missing a trick to make this better.

Additionally, callers would need to perform a caps check against `Client.BuildOpts().LLBCaps`, which I don't *think* is possible to push into this function call, since we need access to a client/caps object.

Reply:
Suggestion: we can work around the caps issue by attaching the `NewResult` to the client struct, and doing caps resolution there.

Reply:
Another way is to do it in https://github.com/moby/buildkit/blob/master/frontend/gateway/grpcclient/client.go#L114 and filter out attestations if the cap is not present.

Reply:
Have added a new gateway capability for this, and done your suggestion.

I think we might be able to rework the `AddInTotoAttestation` back to a generic `AddAttestation` with the #3036 proposal, though I'm not too worried about that.